# 🧠 Notebook 00 — Project Scope & Reproducibility

---

<div style="background: linear-gradient(135deg, #0f0c29 0%, #302b63 50%, #24243e 100%); padding: 24px 28px; border-radius: 16px; margin: 12px 0; border: 1px solid #e94560;">
<h2 style="color: #e94560; margin: 0 0 8px 0; font-size: 1.4em;">⚡ Peramalan Penjualan Produk Mie Instan</h2>
<h3 style="color: #a8dadc; margin: 0 0 16px 0; font-weight: 400; font-size: 1.05em;">Model Hibrida CNN–BiLSTM dengan Optimasi Genetic Algorithm</h3>
<hr style="border: none; border-top: 1px solid #444; margin: 12px 0;">
<table style="color: #ccc; font-size: 0.9em; border-collapse: collapse; width: 100%;">
<tr><td style="padding: 4px 16px 4px 0; white-space:nowrap;">📍 Studi Kasus</td><td><strong style="color: #fff;">PT. XYZ Banjarmasin, Kalimantan Selatan</strong></td></tr>
<tr><td style="padding: 4px 16px 4px 0;">🎓 Mata Kuliah</td><td><strong style="color: #fff;">Tugas Akhir — ES234733</strong></td></tr>
<tr><td style="padding: 4px 16px 4px 0;">🏛 Institusi</td><td><strong style="color: #fff;">Departemen Sistem Informasi, FTEIC — ITS Surabaya</strong></td></tr>
<tr><td style="padding: 4px 16px 4px 0;">👤 Penulis</td><td><strong style="color: #fff;">Muhammad Iqbal Baiduri Yamani — NRP 5026221103</strong></td></tr>
<tr><td style="padding: 4px 16px 4px 0;">🧑‍🏫 Pembimbing</td><td><strong style="color: #fff;">Edwin Riksakomara, S.Kom., M.T.</strong></td></tr>
<tr><td style="padding: 4px 16px 4px 0;">📅 Tahun</td><td><strong style="color: #fff;">2025</strong></td></tr>
</table>
</div>

## 📌 Tujuan Notebook Ini

Notebook 00 adalah **titik awal pipeline eksperimen**. Seluruh notebook (01–18) bergantung pada konfigurasi yang ditetapkan di sini.

| # | Tahapan | Keterangan |
|---|---------|------------|
| 1 | **Import Library & Verifikasi Versi** | Memastikan seluruh dependency tersedia dengan versi yang terdokumentasi |
| 2 | **Global Seed Lock** | Mengunci semua random seed agar eksperimen dapat direproduksi |
| 3 | **Deteksi Hardware** | Mencatat spesifikasi CPU/GPU yang digunakan |
| 4 | **Struktur Folder Output** | Membuat direktori hasil secara terprogram |
| 5 | **Metadata Eksperimen** | Menyimpan log JSON yang menjadi referensi seluruh pipeline |
| 6 | **Reproducibility Checklist** | Verifikasi final semua item selesai |

---

### 🔬 Konteks Penelitian

**Domain:** Industri FMCG (*Fast Moving Consumer Goods*) — aliran produk dari **produsen ke distributor**.

**Masalah:** Peramalan permintaan mingguan produk mie instan yang memiliki pola non-linier dan ketergantungan temporal kompleks, di mana kesalahan peramalan berdampak langsung pada *stockout* atau *overstock*.

**Solusi yang Diusulkan (arsitektur model):**

```mermaid
flowchart LR
    A[Input Mingguan<br/>tau = 8] --> B[Conv1D] --> C[MaxPool1D] --> D[Flatten] --> E[BiLSTM] --> F[Dense Output] --> G[Prediksi Y t+1]
    H[Genetic Algorithm] -. optimasi 7 hyperparameter .-> B
    H -. optimasi 7 hyperparameter .-> C
    H -. optimasi 7 hyperparameter .-> E
```

| Aspek | Detail |
|-------|--------|
| **Data** | Penjualan mingguan 2019–2025, ~364 observasi, Top-1 SKU mie instan |
| **Target** | Grand Total (karton/minggu) |
| **Sliding Window** | τ = 8 minggu → 283 pasang latih (setelah split 80:20) |
| **Split Ratios** | 60:40 · 70:30 · **80:20** · 90:10 |
| **Validasi** | Walk-Forward CV k=5, *expanding window* |
| **Metrik Utama** | MAPE (primer), MAE, RMSE, R² (pendamping) |
| **GA Hyperparameters** | `filters`, `kernel_size`, `pool_size`, `bilstm_units`, `dropout_rate`, `learning_rate`, `batch_size` |

## 🗺️ Peta Pipeline End-to-End

### Visual Flow (Rapi & Terstruktur)

```mermaid
flowchart LR
    subgraph A[FASE A - Fondasi Data dan Eksperimen]
        N00[00 Scope]:::active --> N01[01 Load] --> N02[02 Audit] --> N03[03 EDA] --> N04[04 IQR] --> N05[05 Windowing] --> N06[06 Split] --> N07[07 Walk-Forward CV]
    end

    subgraph B[FASE B - Eksperimen Deep Learning]
        N08[08-13 CNN BiLSTM CNN-BiLSTM + GA]
    end

    subgraph C[FASE C - Baseline dan Komparasi]
        N14[14-15 Baseline Klasik + Perbandingan]
    end

    subgraph D[FASE D - Packaging Laporan]
        N16[16-17 Asset BAB 4 dan BAB 5]
    end

    subgraph E[FASE E - Audit Akhir]
        N18[18 Leakage dan Reproducibility Audit]
    end

    N07 --> N08 --> N14 --> N16 --> N18

    classDef active fill:#ffe082,stroke:#ef6c00,color:#1f1f1f,stroke-width:2px;
```

### Keterangan Fase A (00-07)

| Kode | Fokus |
|------|-------|
| 00 | Setup reproducibility, metadata, dan folder output |
| 01 | Baca dataset dan validasi schema |
| 02 | Audit integritas index waktu mingguan |
| 03 | EDA univariat dan pola musiman |
| 04 | Penanganan outlier dengan IQR clipping |
| 05 | Framing supervised dengan sliding window (tau = 8) |
| 06 | Builder split 60:40, 70:30, 80:20, 90:10 |
| 07 | Walk-forward cross validation (k = 5) |

### Ringkasan Fase Lanjutan

| Fase | Cakupan | Fokus |
|------|---------|-------|
| B | 08-13 | Eksperimen Deep Learning (CNN, BiLSTM, CNN-BiLSTM; baseline + GA) |
| C | 14-15 | Baseline klasik + komparasi komprehensif |
| D | 16-17 | Packaging aset BAB 4 dan BAB 5 |
| E | 18 | Audit final leakage dan reproducibility |

---

## 📦 1. Import Library & Verifikasi Versi Environment

> Seluruh library yang dibutuhkan sepanjang pipeline diimpor dan versinya dicatat.  
> Pastikan semua dependency terinstall sebelum menjalankan pipeline.

In [1]:

# ─────────────────────────────────────────────────────────────
#  Verifikasi Ketersediaan Package
#  Semua package sudah terinstall via: uv pip install
#  Tidak perlu re-install — cell ini hanya memverifikasi.
# ─────────────────────────────────────────────────────────────
import importlib.util
import sys

REQUIRED_PACKAGES = {
    "numpy"       : "numpy",
    "pandas"      : "pandas",
    "matplotlib"  : "matplotlib",
    "seaborn"     : "seaborn",
    "sklearn"     : "scikit-learn",
    "openpyxl"    : "openpyxl",
    "tensorflow"  : "tensorflow",
    "keras"       : "keras",
}

print("=" * 60)
print("  VERIFIKASI PACKAGE")
print("=" * 60)
print(f"  Python  : {sys.executable}")
print("-" * 60)

all_ok = True
for import_name, pkg_name in REQUIRED_PACKAGES.items():
    found = importlib.util.find_spec(import_name) is not None
    status = "✅ FOUND" if found else "❌ MISSING"
    if not found:
        all_ok = False
    print(f"  {pkg_name:<16} → {status}")

print("=" * 60)
if all_ok:
    print("  ✅ Semua package tersedia. Lanjut ke cell berikutnya.")
else:
    print("  ❌ Ada package yang tidak ditemukan.")
    print("     Jalankan di terminal:")
    print("     uv pip install --python .venv312\\Scripts\\python.exe \\")
    print("         numpy pandas matplotlib seaborn scikit-learn \\")
    print("         openpyxl tensorflow ipykernel")
print("=" * 60)


  VERIFIKASI PACKAGE
  Python  : c:\Users\mikba\Downloads\Semester 7\Pra Tugas Akhir (Retno Aulia Vinarti)\Repository\CNN-BiLSTM-GA-Sales-Forecasting\.venv312\Scripts\python.exe
------------------------------------------------------------
  numpy            → ✅ FOUND
  pandas           → ✅ FOUND
  matplotlib       → ✅ FOUND
  seaborn          → ✅ FOUND
  scikit-learn     → ✅ FOUND
  openpyxl         → ✅ FOUND
  tensorflow       → ✅ FOUND
  keras            → ✅ FOUND
  ✅ Semua package tersedia. Lanjut ke cell berikutnya.


In [2]:
# ─────────────────────────────────────────────────────────────
#  Notebook 00 — Project Scope & Reproducibility
#  CNN-BiLSTM + GA Sales Forecasting · PT. XYZ Banjarmasin
# ─────────────────────────────────────────────────────────────

import sys
import os
import platform
import warnings
import json
import multiprocessing
import random
import importlib
from datetime import datetime
from pathlib import Path

# Suppress noisy TF logs sebelum import
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3'
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd
import matplotlib
import seaborn as sns
import openpyxl
from sklearn import __version__ as sklearn_version

try:
    tf = importlib.import_module("tensorflow")
except ModuleNotFoundError as exc:
    raise ModuleNotFoundError(
        "TensorFlow belum tersedia di interpreter ini. "
        "Gunakan Python 3.11/3.12 environment untuk menjalankan notebook ini."
    ) from exc

# ── Cetak versi semua library ─────────────────────────────────
print("=" * 60)
print("  VERIFIKASI VERSI ENVIRONMENT")
print("=" * 60)
print(f"  Python          : {sys.version.split()[0]}")
print(f"  NumPy           : {np.__version__}")
print(f"  Pandas          : {pd.__version__}")
print(f"  Matplotlib      : {matplotlib.__version__}")
print(f"  Seaborn         : {sns.__version__}")
print(f"  Scikit-learn    : {sklearn_version}")
print(f"  TensorFlow      : {tf.__version__}")
print(f"  Keras (via TF)  : {tf.keras.__version__}")
print(f"  OpenPyXL        : {openpyxl.__version__}")
print(f"  Platform        : {platform.platform()}")
print("=" * 60)
print(f"  Tanggal Run     : {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
print("=" * 60)

  VERIFIKASI VERSI ENVIRONMENT
  Python          : 3.12.13
  NumPy           : 2.4.4
  Pandas          : 3.0.2
  Matplotlib      : 3.10.9
  Seaborn         : 0.13.2
  Scikit-learn    : 1.8.0
  TensorFlow      : 2.21.0
  Keras (via TF)  : 3.14.0
  OpenPyXL        : 3.1.5
  Platform        : Windows-11-10.0.26200-SP0
  Tanggal Run     : 2026-05-06 05:20:15


## 🌱 2. Global Seed Lock — Reproducibility

> **Mengapa seed harus dikunci?**  
> Deep learning melibatkan banyak sumber keacakan: inisialisasi bobot, dropout, shuffle mini-batch, dan operasi CUDA.  
> Tanpa seed yang konsisten, hasil eksperimen tidak dapat direproduksi — bahkan di mesin yang sama.
>
> Pipeline ini mengunci **5 lapisan** sumber keacakan:
>
> | Layer | Target |
> |-------|--------|
> | `PYTHONHASHSEED` | Hash built-in Python |
> | `random.seed()` | Modul random standar |
> | `numpy.random.seed()` | Operasi NumPy |
> | `tf.random.set_seed()` | Operasi TensorFlow/Keras |
> | `tf.config.experimental.enable_op_determinism()` | CUDA op level |
>
> **Nilai seed:** `GLOBAL_SEED = 42` (digunakan konsisten di seluruh pipeline)

In [3]:
# ── Konfigurasi GLOBAL_SEED ───────────────────────────────────
GLOBAL_SEED = 42

# Layer 1: Python hash seed
os.environ['PYTHONHASHSEED'] = str(GLOBAL_SEED)

# Layer 2: Python built-in random
random.seed(GLOBAL_SEED)

# Layer 3: NumPy
np.random.seed(GLOBAL_SEED)

# Layer 4: TensorFlow / Keras
tf.random.set_seed(GLOBAL_SEED)

# Layer 5: TF deterministic ops (GPU level)
try:
    tf.config.experimental.enable_op_determinism()
    determinism_status = "✅ ENABLED"
except Exception:
    # Tersedia sejak TF 2.9; older versions: partially deterministic
    determinism_status = "⚠️  Partial (TF < 2.9, upgrade recommended)"

# ── Konfirmasi ────────────────────────────────────────────────
print("=" * 60)
print("  GLOBAL SEED LOCK — REPRODUCIBILITY TERJAMIN")
print("=" * 60)
print(f"  GLOBAL_SEED              : {GLOBAL_SEED}")
print(f"  PYTHONHASHSEED           : {os.environ['PYTHONHASHSEED']}  ✅")
print(f"  random.seed()            : ✅ SET")
print(f"  numpy.random.seed()      : ✅ SET")
print(f"  tf.random.set_seed()     : ✅ SET")
print(f"  TF Op Determinism        : {determinism_status}")
print("=" * 60)

# Verifikasi konsistensi numerik
np.random.seed(GLOBAL_SEED)  # reset untuk verifikasi
verify_sample = np.random.rand(3)
print(f"  Verifikasi NumPy random  : {verify_sample}")
print(f"  (nilai di atas harus konsisten di setiap run)")
print("=" * 60)

  GLOBAL SEED LOCK — REPRODUCIBILITY TERJAMIN
  GLOBAL_SEED              : 42
  PYTHONHASHSEED           : 42  ✅
  random.seed()            : ✅ SET
  numpy.random.seed()      : ✅ SET
  tf.random.set_seed()     : ✅ SET
  TF Op Determinism        : ✅ ENABLED
  Verifikasi NumPy random  : [0.37454012 0.95071431 0.73199394]
  (nilai di atas harus konsisten di setiap run)


## 💻 3. Deteksi Hardware

> Spesifikasi hardware dicatat sebagai bagian dari metadata eksperimen.  
> Jika GPU tersedia, **memory growth** diaktifkan untuk mencegah alokasi memori penuh sekaligus.

In [4]:
# ── CPU Info ──────────────────────────────────────────────────
cpu_count   = multiprocessing.cpu_count()
cpu_info    = platform.processor() or "N/A"

# ── GPU Info ──────────────────────────────────────────────────
gpus = tf.config.list_physical_devices('GPU')
gpu_details = []

if gpus:
    for gpu in gpus:
        try:
            details = tf.config.experimental.get_device_details(gpu)
            gpu_details.append(f"  → {gpu.name} — {details.get('device_name', 'Unknown')}")
        except Exception:
            gpu_details.append(f"  → {gpu.name}")
    # Enable memory growth (hindari OOM)
    for gpu in gpus:
        try:
            tf.config.experimental.set_memory_growth(gpu, True)
        except RuntimeError:
            pass
    gpu_mode   = f"{len(gpus)} GPU — Memory Growth ENABLED"
    gpu_note   = ""
else:
    gpu_details = ["  GPU Mode            : CPU-only mode (menggunakan TF CPU backend)"]
    gpu_mode    = "CPU-only mode (menggunakan TF CPU backend)"
    gpu_note    = (
        "\n  ℹ️  Catatan: Model akan berjalan di CPU.\n"
        "     Training GA mungkin memerlukan waktu lebih lama.\n"
        "     Pertimbangkan Google Colab (T4 GPU) untuk training GA."
    )

# ── Print ─────────────────────────────────────────────────────
print("=" * 60)
print("  DETEKSI HARDWARE")
print("=" * 60)
print(f"  CPU Cores (Logical) : {cpu_count}")
print(f"  CPU Processor       : {cpu_info}")
print(f"  RAM Total (approx)  : Cek via psutil (opsional)")
print(f"  GPU Terdeteksi      : {len(gpus)} unit")
for gd in gpu_details:
    print(gd)
print(f"  TF Built with CUDA  : {tf.test.is_built_with_cuda()}")
print(f"  GPU Available       : {len(gpus) > 0}")
if gpu_note:
    print("-" * 60 + gpu_note)
print("=" * 60)

# ── Simpan ke dict global ─────────────────────────────────────
HARDWARE_INFO = {
    "cpu_count"       : cpu_count,
    "cpu_info"        : cpu_info,
    "gpu_count"       : len(gpus),
    "gpu_names"       : [g.name for g in gpus],
    "tf_cuda_support" : bool(tf.test.is_built_with_cuda()),
    "gpu_available"   : len(gpus) > 0,
    "compute_mode"    : "GPU" if gpus else "CPU"
}

  DETEKSI HARDWARE
  CPU Cores (Logical) : 20
  CPU Processor       : Intel64 Family 6 Model 186 Stepping 2, GenuineIntel
  RAM Total (approx)  : Cek via psutil (opsional)
  GPU Terdeteksi      : 0 unit
  GPU Mode            : CPU-only mode (menggunakan TF CPU backend)
  TF Built with CUDA  : False
  GPU Available       : False
------------------------------------------------------------
  ℹ️  Catatan: Model akan berjalan di CPU.
     Training GA mungkin memerlukan waktu lebih lama.
     Pertimbangkan Google Colab (T4 GPU) untuk training GA.


## 📁 4. Struktur Folder Output

> Seluruh artefak eksperimen (model, metrik, gambar, fold CV) disimpan dalam struktur direktori terorganisir.  
> Folder dibuat otomatis; jika sudah ada, tidak ada perubahan.

```
CNN-BiLSTM-GA-Sales-Forecasting/
├── data/                     ← Dataset mentah (sudah ada)
├── notebook/                 ← Semua notebook 00–18
├── outputs/
│   ├── figures/              ← Semua plot dan visualisasi (EDA, training, evaluasi)
│   ├── models/               ← File .keras model terbaik per split
│   ├── metrics/              ← Tabel MAPE, MAE, RMSE, R² per model per split
│   ├── splits/               ← X_train, X_test, y_train, y_test per rasio
│   ├── cv_folds/             ← Fold walk-forward CV (NB07)
│   ├── checkpoints/          ← Checkpoint training sementara
│   ├── reports/              ← Laporan komprehensif per split
│   └── assets/               ← Asset pendukung (scaler, encoder, dll)
└── logs/                     ← experiment_metadata.json, audit log
```


In [5]:
# ── Root Directory ────────────────────────────────────────────
# Notebook berada di ./notebook/ → root satu level di atas
ROOT_DIR = Path(".").resolve().parent

# ── Definisi Folder ───────────────────────────────────────────
FOLDER_SPEC = [
    ("outputs",              ROOT_DIR / "outputs"),
    ("outputs/figures",      ROOT_DIR / "outputs" / "figures"),
    ("outputs/models",       ROOT_DIR / "outputs" / "models"),
    ("outputs/metrics",      ROOT_DIR / "outputs" / "metrics"),
    ("outputs/splits",       ROOT_DIR / "outputs" / "splits"),
    ("outputs/cv_folds",     ROOT_DIR / "outputs" / "cv_folds"),
    ("outputs/checkpoints",  ROOT_DIR / "outputs" / "checkpoints"),
    ("outputs/reports",      ROOT_DIR / "outputs" / "reports"),
    ("outputs/assets",       ROOT_DIR / "outputs" / "assets"),
    ("logs",                 ROOT_DIR / "logs"),
]

# ── Buat Folder ───────────────────────────────────────────────
print("=" * 60)
print("  STRUKTUR FOLDER OUTPUT")
print("=" * 60)

created_count = 0
for name, path in FOLDER_SPEC:
    existed_before = path.exists()
    path.mkdir(parents=True, exist_ok=True)
    status = "✅ OK (sudah ada)" if existed_before else "✅ CREATED"
    indent = "    " if "/" in name else "  "
    label  = f"{indent}📁 {name:<24}"
    print(f"{label}→ {status}")
    created_count += 1

print("=" * 60)
print(f"  Root   : .../{ROOT_DIR.name}")
print(f"  Total  : {created_count} folder dibuat/diverifikasi")
print("=" * 60)

# ── Simpan path ke variabel untuk digunakan notebook lain ────
PATHS = {
    "root"         : ROOT_DIR,
    "data"         : ROOT_DIR / "data",
    "logs"         : ROOT_DIR / "logs",
    "outputs"      : ROOT_DIR / "outputs",
    "figures"      : ROOT_DIR / "outputs" / "figures",
    "models"       : ROOT_DIR / "outputs" / "models",
    "metrics"      : ROOT_DIR / "outputs" / "metrics",
    "splits"       : ROOT_DIR / "outputs" / "splits",
    "cv_folds"     : ROOT_DIR / "outputs" / "cv_folds",
    "checkpoints"  : ROOT_DIR / "outputs" / "checkpoints",
    "reports"      : ROOT_DIR / "outputs" / "reports",
    "assets"       : ROOT_DIR / "outputs" / "assets",
    "notebooks"    : ROOT_DIR / "notebook",
    "raw_data_file": ROOT_DIR / "data" / "Dataset_Raw_Top 1 Sales 2019 - 2025.xlsx",
}


  STRUKTUR FOLDER OUTPUT
  📁 outputs                 → ✅ OK (sudah ada)
    📁 outputs/figures         → ✅ OK (sudah ada)
    📁 outputs/models          → ✅ OK (sudah ada)
    📁 outputs/metrics         → ✅ OK (sudah ada)
    📁 outputs/splits          → ✅ OK (sudah ada)
    📁 outputs/cv_folds        → ✅ OK (sudah ada)
    📁 outputs/checkpoints     → ✅ OK (sudah ada)
    📁 outputs/reports         → ✅ OK (sudah ada)
    📁 outputs/assets          → ✅ OK (sudah ada)
  📁 logs                    → ✅ OK (sudah ada)
  Root   : .../CNN-BiLSTM-GA-Sales-Forecasting
  Total  : 10 folder dibuat/diverifikasi


## 📋 5. Metadata Eksperimen

> Seluruh konfigurasi eksperimen disimpan dalam satu file JSON: `logs/experiment_metadata.json`.  
> File ini menjadi **referensi tunggal** (single source of truth) yang dapat dimuat oleh notebook manapun
> untuk memverifikasi bahwa eksperimen berjalan dengan konfigurasi yang benar.

In [ ]:
# ── Susun Metadata Lengkap ────────────────────────────────────
EXPERIMENT_METADATA = {
    "project": {
        "title"       : (
            "Peramalan Penjualan Produk Mie Instan Menggunakan "
            "Model Hibrida CNN-BiLSTM dengan Optimasi Genetic Algorithm"
        ),
        "title_en"    : (
            "Instant Noodle Sales Forecasting Using Hybrid CNN-BiLSTM "
            "Model Optimized with Genetic Algorithm"
        ),
        "case_study"  : "PT. XYZ Banjarmasin",
        "course_code" : "ES234733",
        "institution" : "Institut Teknologi Sepuluh Nopember (ITS) Surabaya",
        "department"  : "Departemen Sistem Informasi, FTEIC",
    },
    "author": {
        "name"           : "Muhammad Iqbal Baiduri Yamani",
        "nrp"            : "5026221103",
        "supervisor"     : "Edwin Riksakomara, S.Kom., M.T.",
        "supervisor_nip" : "196907252003121001",
    },
    "experiment": {
        "global_seed"       : GLOBAL_SEED,
        "run_date"          : datetime.now().strftime("%Y-%m-%d %H:%M:%S"),
        "python_version"    : sys.version.split()[0],
        "tensorflow_version": tf.__version__,
        "numpy_version"     : np.__version__,
        "pandas_version"    : pd.__version__,
        "sklearn_version"   : sklearn_version,
        "platform"          : platform.platform(),
    },
    "data": {
        "domain"          : "FMCG Weekly Sales (Manufacturer to Distributor)",
        "product"         : "Mie Instan — Top-1 SKU by Revenue",
        "granularity"     : "Weekly",
        "period"          : "2019–2025",
        "target_column"   : "Grand Total",
        "n_weeks_approx"  : 364,
        "file_name"       : "Dataset_Raw_Top 1 Sales 2019 - 2025.xlsx",
    },
    "methodology": {
        "model"              : "CNN-BiLSTM + GA",
        "architecture"       : "Conv1D → MaxPool1D → Flatten → BiLSTM → Dropout → Dense",
        "sliding_window_tau" : 8,
        "forecast_horizon"   : 1,
        "outlier_handling"   : "IQR Clipping (Q1 - 1.5*IQR, Q3 + 1.5*IQR)",
        "splits"             : ["60:40", "70:30", "80:20", "90:10"],
        "validation"         : "Walk-Forward CV (k=5, expanding window)",
        "ga_hyperparameters" : [
            "filters", "kernel_size", "pool_size",
            "bilstm_units", "dropout_rate", "learning_rate", "batch_size"
        ],
        "metrics"            : ["MAPE", "MAE", "RMSE", "R²"],
        "primary_metric"     : "MAPE",
        "comparison_models"  : [
            "CNN", "CNN+GA", "BiLSTM", "BiLSTM+GA",
            "CNN-BiLSTM", "CNN-BiLSTM+GA",
            "Naive", "Moving Average", "ARIMA", "Prophet", "XGBoost"
        ],
    },
    "hardware"    : HARDWARE_INFO,
    "pipeline_version": "1.0.0",
    "notebook"    : "00 - Project Scope and Reproducibility",
}

# ── Simpan ke JSON ────────────────────────────────────────────
metadata_path = PATHS["logs"] / "experiment_metadata.json"
with open(metadata_path, "w", encoding="utf-8") as f:
    json.dump(EXPERIMENT_METADATA, f, indent=2, ensure_ascii=False)

# ── Print Ringkasan ───────────────────────────────────────────
meta = EXPERIMENT_METADATA
print("=" * 60)
print("  METADATA EKSPERIMEN TERSIMPAN")
print("=" * 60)
print(f"  📄 File  : logs/experiment_metadata.json")
print()
print("  ── Project " + "─" * 46)
print(f"  title        : {meta['project']['title'][:40]}...")
print(f"  case_study   : {meta['project']['case_study']}")
print(f"  course_code  : {meta['project']['course_code']}")
print(f"  institution  : Institut Teknologi Sepuluh Nopember (ITS)")
print()
print("  ── Author " + "─" * 47)
print(f"  name         : {meta['author']['name']}")
print(f"  nrp          : {meta['author']['nrp']}")
print(f"  supervisor   : {meta['author']['supervisor']}")
print()
print("  ── Experiment Config " + "─" * 36)
print(f"  global_seed  : {meta['experiment']['global_seed']}")
print(f"  run_date     : {meta['experiment']['run_date']}")
print(f"  tf_version   : {meta['experiment']['tensorflow_version']}")
print(f"  numpy        : {meta['experiment']['numpy_version']}")
print(f"  python       : {meta['experiment']['python_version']}")
print()
print("  ── Methodology " + "─" * 42)
mth = meta['methodology']
print(f"  model        : {mth['model']}")
print(f"  tau          : {mth['sliding_window_tau']}")
print(f"  horizon      : {mth['forecast_horizon']}")
print(f"  splits       : {mth['splits']}")
print(f"  validation   : {mth['validation']}")
gp = mth['ga_hyperparameters']
print(f"  ga_params    : {gp[:4]}")
print(f"                  {gp[4:]}")
print(f"  metrics      : {mth['metrics']}")
print(f"  primary      : {mth['primary_metric']}")
print()
print("  ── Hardware " + "─" * 45)
print(f"  compute_mode : {HARDWARE_INFO['compute_mode']}")
print(f"  cpu_count    : {HARDWARE_INFO['cpu_count']}")
print(f"  gpu_count    : {HARDWARE_INFO['gpu_count']}")
print("=" * 60)
print(f"  ✅ Metadata berhasil disimpan ke logs/experiment_metadata.json")
print("=" * 60)

  METADATA EKSPERIMEN TERSIMPAN
  📄 File  : logs/experiment_metadata.json

  ── Project ──────────────────────────────────────────────
  title        : Peramalan Penjualan Produk Mie Instan Me...
  case_study   : PT. XYZ Banjarmasin
  course_code  : ES234733
  institution  : Institut Teknologi Sepuluh Nopember (ITS)

  ── Author ───────────────────────────────────────────────
  name         : Muhammad Iqbal Baiduri Yamani
  nrp          : 5026221103
  supervisor   : Edwin Riksakomara, S.Kom., M.T.

  ── Experiment Config ────────────────────────────────────
  global_seed  : 42
  run_date     : 2026-05-06 05:21:14
  tf_version   : 2.21.0
  numpy        : 2.4.4
  python       : 3.12.13

  ── Methodology ──────────────────────────────────────────
  model        : CNN-BiLSTM + GA
  tau          : 8
  horizon      : 1
  splits       : ['60:40', '70:30', '80:20', '90:10']
  validation   : Walk-Forward CV (k=5, expanding window)
  ga_params    : ['filters', 'kernel_size', 'pool_size', 'bilstm

## ✅ 6. Reproducibility Checklist

> Verifikasi bahwa semua persyaratan reproducibility dari Pipeline.md telah dipenuhi  
> sebelum melanjutkan ke Notebook 01.

| # | Item | Status |
|---|------|--------|
| 1 | Seed global ditetapkan | ✅ `GLOBAL_SEED = 42` dikunci di 5 layer |
| 2 | Versi environment dicatat | ✅ TF 2.21, NumPy 2.4, Python 3.12 |
| 3 | Struktur output hasil dibuat | ✅ 10 folder tersedia |
| 4 | Format log eksperimen ditetapkan | ✅ `logs/experiment_metadata.json` v1.0.0 |

In [7]:
# ── Reproducibility Checklist ─────────────────────────────────
checklist = [
    (
        "Seed global ditetapkan",
        f"GLOBAL_SEED = {GLOBAL_SEED} (Python, NumPy, TensorFlow)"
    ),
    (
        "Versi environment dicatat",
        f"TF={tf.__version__}, NumPy={np.__version__}, Python={sys.version.split()[0]}"
    ),
    (
        "Struktur output folder dibuat",
        "outputs/(figures, models, metrics, splits, cv_folds, checkpoints, reports, assets), logs/"
    ),
    (
        "Format log eksperimen ditetapkan",
        "logs/experiment_metadata.json (schema v1.0.0)"
    ),
    (
        "Hardware info tercatat",
        f"CPU={HARDWARE_INFO['cpu_count']} cores, "
        f"GPU={HARDWARE_INFO['gpu_count']}, "
        f"Compute={HARDWARE_INFO['compute_mode']}"
    ),
    (
        "Path artefak terdefinisi",
        "PATHS dict tersedia untuk seluruh pipeline"
    ),
]

print("=" * 60)
print("  REPRODUCIBILITY CHECKLIST — NOTEBOOK 00")
print("=" * 60)

all_pass = True
for item, detail in checklist:
    print(f"\n  ✅  {item}")
    print(f"      → {detail}")

print()
print("=" * 60)
status = "✅ SEMUA ITEM SELESAI — NOTEBOOK 00 COMPLETE" if all_pass else "❌ ADA ITEM BELUM SELESAI"
print(f"  STATUS   : {status}")
print("=" * 60)
print()
print("  🚀 Lanjut ke:")
print("     📓 Notebook 01 — Data Load and Schema Validation")
print()

  REPRODUCIBILITY CHECKLIST — NOTEBOOK 00

  ✅  Seed global ditetapkan
      → GLOBAL_SEED = 42 (Python, NumPy, TensorFlow)

  ✅  Versi environment dicatat
      → TF=2.21.0, NumPy=2.4.4, Python=3.12.13

  ✅  Struktur output folder dibuat
      → outputs/(figures, models, metrics, splits, cv_folds, checkpoints, reports, assets), logs/

  ✅  Format log eksperimen ditetapkan
      → logs/experiment_metadata.json (schema v1.0.0)

  ✅  Hardware info tercatat
      → CPU=20 cores, GPU=0, Compute=CPU

  ✅  Path artefak terdefinisi
      → PATHS dict tersedia untuk seluruh pipeline

  STATUS   : ✅ SEMUA ITEM SELESAI — NOTEBOOK 00 COMPLETE

  🚀 Lanjut ke:
     📓 Notebook 01 — Data Load and Schema Validation



---

## 🔗 Navigasi Pipeline

| | Notebook |
|--|----------|
| ← | *(tidak ada notebook sebelumnya)* |
| **→** | **[01 - Data Load and Schema Validation](./01%20-%20Data%20Load%20and%20Schema%20Validation.ipynb)** |

---

### 📎 Variabel Penting yang Ditetapkan di Notebook Ini

Variabel berikut perlu didefinisikan ulang (atau dimuat dari metadata JSON) di notebook lain:

```python
# ─── Salin blok ini ke bagian atas notebook lain ───
import os, json, random
from pathlib import Path
import numpy as np
import tensorflow as tf

GLOBAL_SEED = 42
os.environ['PYTHONHASHSEED'] = str(GLOBAL_SEED)
random.seed(GLOBAL_SEED)
np.random.seed(GLOBAL_SEED)
tf.random.set_seed(GLOBAL_SEED)

ROOT_DIR = Path(".").resolve().parent
PATHS = {
    "root"         : ROOT_DIR,
    "data"         : ROOT_DIR / "data",
    "logs"         : ROOT_DIR / "logs",
    "outputs"      : ROOT_DIR / "outputs",
    "figures"      : ROOT_DIR / "outputs" / "figures",
    "models"       : ROOT_DIR / "outputs" / "models",
    "metrics"      : ROOT_DIR / "outputs" / "metrics",
    "splits"       : ROOT_DIR / "outputs" / "splits",
    "cv_folds"     : ROOT_DIR / "outputs" / "cv_folds",
    "checkpoints"  : ROOT_DIR / "outputs" / "checkpoints",
    "reports"      : ROOT_DIR / "outputs" / "reports",
    "assets"       : ROOT_DIR / "outputs" / "assets",
    "notebooks"    : ROOT_DIR / "notebook",
    "raw_data_file": ROOT_DIR / "data" / "Dataset_Raw_Top 1 Sales 2019 - 2025.xlsx",
}
```

---

<div style="text-align: center; color: #888; font-size: 0.85em; padding: 12px 0;">
Notebook 00 — Project Scope &amp; Reproducibility &nbsp;|&nbsp;
CNN–BiLSTM + GA Sales Forecasting &nbsp;|&nbsp;
Departemen Sistem Informasi ITS Surabaya &nbsp;|&nbsp; 2025
</div>